File: Task1.cpp
#include <iostream>  
#include <cmath>  
#include <vector>  
  
using namespace std;  
  
int NOD (int a, int b) {  
    if (b == 0) {  
        return a;  
    }  
    return NOD(b , a % b);  
}  
  
bool Prime(int a) {  
    if (a == 1)  return false;  
    for (int i = 2; i < sqrt(a); i++) {  
        if (a % i == 0) return false;  
    }  
    return true;  
}  
  
int EilerFunc(int p) { // определяем количество взаимно простых чисел для p   
    int result = p;  
    for (int i = 2; i <= sqrt(p); i++) { //проходимся до корня из числа  
        //если число разделилось на i, то мы уменьшаем р на i так как все числа,  
        while (p % i == 0) { // которые делятся на i не являются взаимно простыми с p  
            p /= i;  
        }  
        result -= result / i; // из результата вычитаем количество чисел result/i  
    }  
    if (p > 1) { // если п - простое число, которое не разделилось, то оно простое и результатом  
        result -= result / p; // функции будет р-1  
    }  
  
    return result;  
}  
  
int modexp(int basis, int power, int modP) {  
    if (power == 0) {  
        return 1;  
    }  
    int z = modexp(basis % modP, power / 2, modP) % modP;  
    if (power % 2 == 0) {  
        return (z * z) % modP;  
    }  
    else {  
        return ((basis % modP) * ((z * z) % modP)) % modP;  
    }  
}  
  
pair<int, int> remainder(int& basis, int& power, int& modP) {  
    pair<int, int> ost = { 1,1 };  
  
    ost.first = modexp(basis, power, modP);  
  
    int logByNumber = log2(power);     //рассматриваем остатки через логарифм и 2сс.  
    vector<int> binary;  //переводим степень в 2сс.  
    int power2 = power;  
    while (power2 > 0) {  
        binary.push_back(power2 % 2);  
        power2 /= 2;  
    }  
  
    int resultForPrime = 1;  
    for (int i = 0; i <= logByNumber; i++) {  
        if (binary[i] == 1) { //бинарное не переворачиваем, а просто идем с "конца"  
            int oneDegree = pow(2, i);  
            resultForPrime *= modexp(basis, oneDegree, modP);  
        }  
    }  
    ost.second = resultForPrime % modP;  //после смотрим на остаток  
  
    return ost;  //вернем пару значений  
} 
  
int main() { 
    
    int basis = 0, basis2 = 0, power = 0, power2 = 0, modP = 0; 
 
    cout << "Введите основание числа: "; 
    cin >> basis; 
    if (cin.fail()) { 
        cout << "Ошибка: введено не число." << endl; 
        return 1; 
    } 
 
    cout << "Введите степень числа: "; 
    cin >> power; 
    if (cin.fail()) { 
        cout << "Ошибка: введено не число." << endl; 
        return 1; 
    } 
 
    cout << "Введите основание второго числа: "; 
    cin >> basis2; 
    if (cin.fail()) { 
        cout << "Ошибка: введено не число." << endl; 
        return 1; 
    } 
 
    cout << "Введите степень числа второго числа: "; 
    cin >> power2; 
    if (cin.fail()) { 
        cout << "Ошибка: введено не число." << endl; 
        return 1; 
    } 
 
    cout << "Введите модуль числа: "; 
    cin >> modP; 
    if (cin.fail()) { 
        cout << "Ошибка: введено не число." << endl; 
        return 1; 
    } 
    
    int result1 = modexp(basis, power, modP); 
    cout << "1-ый результат: " << result1 << endl; 
    int result2 = modexp(basis2, power2, modP); 
    cout << "2-ой результат: " << result2 << endl;  
 
    if (result1 > result2) { 
        cout << basis << "^" << power << " mod " << modP << " > " << basis2 << "^" << power2 << " mod " << modP << endl; 
    } else if (result1 < result2) { 
        cout << basis << "^" << power << " mod " << modP << " < " << basis2 << "^" << power2 << " mod " << modP << endl; 
    } else if (result1 = result2) { 
        cout << basis << "^" << power << " mod " << modP << " = " << basis2 << "^" << power2 << " mod " << modP << endl; 
    } 
    return 0; 
}

File: Task1.cs
using System;
using System.Collections.Generic;

class MainClass {
    static int NOD(int a, int b) {
        if (b == 0) {
            return a;
        }
        return NOD(b, a % b);
    }

    static bool IsPrime(int a) {
        if (a == 1) {
            return false;
        }
        for (int i = 2; i <= Math.Sqrt(a); i++) {
            if (a % i == 0) {
                return false;
            }
        }
        return true;
    }

    static int EilerFunc(int p) {
        int result = p;
        for (int i = 2; i * i <= p; i++) {
            if (p % i == 0) {
                while (p % i == 0) {
                    p /= i;
                }
                result -= result / i;
            }
        }
        if (p > 1) {
            result -= result / p;
        }
        return result;
    }

    static int ModExp(int basis, int power, int modP) {
        if (power == 0) {
            return 1;
        }
        int z = ModExp(basis % modP, power / 2, modP) % modP;
        if (power % 2 == 0) {
            return (z * z) % modP;
        }
        else {
            return ((basis % modP) * ((z * z) % modP)) % modP;
        }
    }

    static Tuple<int, int> Remainder(ref int basis, ref int power, ref int modP) {
        Tuple<int, int> ost = Tuple.Create(1, 1);

        ost = Tuple.Create(ModExp(basis, power, modP), 1);

        int logByNumber = (int)Math.Log(power, 2);
        List<int> binary = new List<int>();
        int power2 = power;
        while (power2 > 0) {
            binary.Add(power2 % 2);
            power2 /= 2;
        }

        int resultForPrime = 1;
        for (int i = 0; i <= logByNumber; i++) {
            if (binary[i] == 1) {
                int oneDegree = (int)Math.Pow(2, i);
                resultForPrime *= ModExp(basis, oneDegree, modP);
            }
        }
        ost = Tuple.Create(ost.Item1, resultForPrime % modP);

        return ost;
    }

    public static void Main(string[] args) {
        int basis = 0, basis2 = 0, power = 0, power2 = 0, modP = 0;
        Console.Write("Введите основание числа: ");
        basis = Convert.ToInt32(Console.ReadLine());
        Console.Write("Введите степень числа: ");
        power = Convert.ToInt32(Console.ReadLine());
        Console.Write("Введите модуль числа: ");
        modP = Convert.ToInt32(Console.ReadLine());

        bool zahod = false;

        if (basis < modP && power == (modP - 1) && IsPrime(modP)) {
            Console.WriteLine($"По теореме Ферма получим: {basis}^{power} mod {modP} = 1");
            var ost = Remainder(ref basis, ref power, ref modP);
            Console.WriteLine($"Проверим, используя логарифм: {ost.Item2}");
            zahod = true;
        }
        if (NOD(basis, modP) == 1 && power == EilerFunc(modP)) {
            zahod = true;
            Console.WriteLine($"По теореме Эйлера получим: {basis}^{power} mod {modP} = 1");
            var ost = Remainder(ref basis, ref power, ref modP);
            Console.WriteLine($"Проверим, используя логарифм: {ost.Item2}");
        }
        if (!zahod) {
            power = EilerFunc(modP);
            int result = ModExp(basis, power, modP);
            var ost = Remainder(ref basis, ref power, ref modP);
            Console.WriteLine($"Результат, используя свойства сравнений = {ost.Item1}");
            Console.WriteLine($"Результат, используя логарифм = {ost.Item2}");
        }

        Console.Write("Введите основание второго числа, которое будете сравнивать по модулю: ");
        basis2 = Convert.ToInt32(Console.ReadLine());
        Console.Write("Введите степень числа второго числа, которое будете сравнивать по модулю: ");
        power2 = Convert.ToInt32(Console.ReadLine());

        if (ModExp(basis, power, modP) > ModExp(basis2, power2, modP)) {
            Console.WriteLine($"{basis}^{power} mod {modP} > {basis2}^{power2} mod {modP}");
        }
        else if (ModExp(basis, power, modP) < ModExp(basis2, power2, modP)) {
            Console.WriteLine($"{basis}^{power} mod {modP} < {basis2}^{power2} mod {modP}");
        }
        else {
            Console.WriteLine($"{basis}^{power} mod {modP} = {basis2}^{power2} mod {modP}");
        }
    }
}
